File: mainActivity/src/main/java/com/sierrawireless/avphone/task/SystemClient.java
package com.sierrawireless.avphone.task;

import android.util.Log;

import com.sierrawireless.avphone.tools.Tools;

import java.io.IOException;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

import net.airvantage.model.AirVantageException;
import net.airvantage.model.Application;
import net.airvantage.model.AvSystem;
import net.airvantage.model.MqttCommunication;
import net.airvantage.utils.AirVantageClient;
import net.airvantage.utils.Utils;

public class SystemClient implements ISystemClient {
    private static final String TAG = "SystemClient";

    private AirVantageClient client;

    public SystemClient(AirVantageClient client) {
        this.client = client;
    }

    @Override
    public net.airvantage.model.AvSystem getSystem(final String serialNumber, String type) throws IOException, AirVantageException {
        List<net.airvantage.model.AvSystem> systems = client.getSystemsBySerialNumber(Tools.buildSerialNumber(serialNumber, type));

        return Utils.firstWhere(systems, AvSystem.hasSerialNumber(Tools.buildSerialNumber(serialNumber, type)));
    }

    @Override
    public net.airvantage.model.AvSystem createSystem(String serialNumber, String iccid, String type, String mqttPassword,
            String applicationUid, String deviceName, String userName, String imei) throws IOException, AirVantageException {
        net.airvantage.model.AvSystem system = new net.airvantage.model.AvSystem();

        Boolean exist = client.getGateway((serialNumber + "-ANDROID-" + type).toUpperCase());
        net.airvantage.model.AvSystem.Gateway gateway = new net.airvantage.model.AvSystem.Gateway();

        if (!exist) {
            gateway.serialNumber = (serialNumber + "-ANDROID-" + type).toUpperCase();
           // gateway.imei = imei + type;
            gateway.type = type;
        }else{
            gateway.serialNumber = (serialNumber + "-ANDROID-" + type).toUpperCase();
        }
        Log.d(TAG, "createSystem: Gateway exit " + exist);
        Log.d(TAG, "gateway is " + (serialNumber + "-ANDROID-" + type).toUpperCase() );
        system.name = deviceName + " de " + userName + "(" + type + ")";

        system.gateway = gateway;

        system.state = "READY";

        Application application = new Application();
        application.uid = applicationUid;
        system.applications = Arrays.asList(application);

        MqttCommunication mqtt = new MqttCommunication();
        mqtt.password = mqttPassword;

        system.communication = new HashMap<String, MqttCommunication>();
        system.communication.put("mqtt", mqtt);
        system.type = type;

        return client.createSystem(system);
    }

    @Override
    public void deleteSystem(net.airvantage.model.AvSystem system)throws IOException, AirVantageException {
        client.deleteSystem(system);
    }

}

File: mainActivity/src/main/java/com/sierrawireless/avphone/task/SystemClient.kt
package com.sierrawireless.avphone.task

import com.sierrawireless.avphone.tools.Tools
import net.airvantage.model.AirVantageException
import net.airvantage.model.Application
import net.airvantage.model.AvSystem
import net.airvantage.model.MqttCommunication
import net.airvantage.utils.AirVantageClient
import net.airvantage.utils.Utils
import java.io.IOException
import java.util.HashMap
import kotlin.collections.ArrayList
import kotlin.collections.set

class SystemClient internal constructor(private val client: AirVantageClient) : ISystemClient {

    @Throws(IOException::class, AirVantageException::class)
    override fun getSystem(serialNumber: String, type: String, deviceName: String): net.airvantage.model.AvSystem? {
        val systems = client.getSystemsBySerialNumber(Tools.buildSerialNumber(serialNumber, type, deviceName))

        return Utils.firstWhere(systems, AvSystem.hasSerialNumber(Tools.buildSerialNumber(serialNumber, type, deviceName)))
    }

    @Throws(IOException::class, AirVantageException::class)
    override fun createSystem(serialNumber: String, iccid: String, type: String, mqttPassword: String,
                              applicationUid: String, deviceName: String, userName: String, imei: String): net.airvantage.model.AvSystem {
        val system = net.airvantage.model.AvSystem()

        val uid = client.getGateway((serialNumber + "-ANDROID-" + type + "-" + deviceName.replace(" ", "_")).toUpperCase())
        val gateway = net.airvantage.model.AvSystem.Gateway()

        if (uid == null) {
            gateway.serialNumber = (serialNumber + "-ANDROID-" + type + "-" +  deviceName.replace(" ", "_")).toUpperCase()
            // gateway.imei = imei + type;
            gateway.type = type
        } else {
            gateway.uid = uid
        }
        system.name = Tools.buildSystemName(deviceName, userName, type)

        system.gateway = gateway

        system.state = "READY"

        val application = Application()
        application.uid = applicationUid
        val tmp = ArrayList<Application>()
        tmp.add(application)
        system.applications = tmp

        val mqtt = MqttCommunication()
        mqtt.password = mqttPassword

        system.communication = HashMap()
        system.communication!!["mqtt"] = mqtt
        system.type = type

        return client.createSystem(system)
    }

    @Throws(IOException::class, AirVantageException::class)
    override fun updateSystem(system: AvSystem, serialNumber: String, iccid: String, type: String, mqttPassword: String,
                              applicationUid: String, deviceName: String, userName: String, imei: String) {


        val uid = client.getGateway((serialNumber + "-ANDROID-" + type + "-" + deviceName.replace(" ", "_")).toUpperCase())
        val gateway = net.airvantage.model.AvSystem.Gateway()

        if (uid == null) {
            gateway.serialNumber = (serialNumber + "-ANDROID-" + type + "-" +  deviceName.replace(" ", "_")).toUpperCase()
            // gateway.imei = imei + type;
            gateway.type = type
        } else {
            gateway.uid = uid
        }
        system.name = Tools.buildSystemName(deviceName, userName, type)

        system.gateway = gateway

        system.state = "READY"

        val application = Application()
        application.uid = applicationUid
        val tmp = ArrayList<Application>()
        tmp.add(application)
        system.applications = tmp

        val mqtt = MqttCommunication()
        mqtt.password = mqttPassword

        system.communication = HashMap()
        system.communication!!["mqtt"] = mqtt
        system.type = type

        client.updateSystem(system)
    }

    @Throws(IOException::class, AirVantageException::class)
    override fun deleteSystem(system: net.airvantage.model.AvSystem) {
        client.deleteSystem(system)
    }
}
